src/System.Text.Primitives/System/Text/Parsing/InvariantSigned.cs

```diff
@@ -899,176 +899,89 @@ public unsafe static bool TryParseInt32(byte* text, int length, out int value, o
-                    // digits. There may be no overflow after Int32OverflowLength if there are leading zeroes.
-                    for (int index = indexOfFirstDigit + 1; index < overflowLength - 1; index++)
+
+                    int firstNonZeroDigitIndex = index;
+                    if (textLength - firstNonZeroDigitIndex < Int32OverflowLength)
                     {
-                        int nextDigit = text[index] - 48; // '0'
-                        if (nextDigit < 0 || nextDigit > 9)
+                        do
                         {
-                            bytesConsumed = index;
-                            value = parsedValue * sign;
-                            return true;
-                        }
-                        parsedValue = parsedValue * 10 + nextDigit;
+                            answer = answer * 10 + num - '0';
+                            index++;
+                            if (index >= textLength) goto Done;
+                            num = text[index];
+                        } while (num >= '0' && num <= '9');
                     }
-                    for (int index = overflowLength - 1; index < text.Length; index++)
+                    else
                     {
-                        int nextDigit = text[index] - 48; // '0'
-                        if (nextDigit < 0 || nextDigit > 9)
+                        do
                         {
-                            bytesConsumed = index;
-                            value = parsedValue * sign;
-                            return true;
-                        }
-                        // If parsedValue > (int.MaxValue / 10), any more appended digits will cause overflow.
-                        // if parsedValue == (int.MaxValue / 10), any nextDigit greater than 7 or 8 (depending on sign) implies overflow.
-                        bool positive = sign > 0;
-                        bool nextDigitTooLarge = nextDigit > 8 || (positive && nextDigit > 7);
-                        if (parsedValue > int.MaxValue / 10 || parsedValue == int.MaxValue / 10 && nextDigitTooLarge)
-                        {
-                            bytesConsumed = 0;
-                            value = default(int);
-                            return false;
-                        }
-                        parsedValue = parsedValue * 10 + nextDigit;
+                            answer = answer * 10 + num - '0';
+                            index++;
+                            if (index - firstNonZeroDigitIndex == Int32OverflowLength - 1)
+                            {
+                                num = text[index];
+                                if (num >= '0' && num <= '9')
+                                {
+                                    num -= (byte)'0';
+                                    if (WillOverFlow(answer, num, sign)) goto FalseExit;
```

Comment:
Here is basically the idea for parsing numbers longer than Int32OverflowLength, I wonder if it will work better:

```C#
        static void Main(string[] args)
        {
            for (long i = int.MaxValue - 1000; i < int.MaxValue + 10L; i++)
            {
                // lets parse "chars" 
                string chars = i.ToString();

                // this will need to come from parsing leading '-'
                int sign = 1;

                // here is our result
                int accumulator = 0;

                foreach (var c in chars)
                {
                    // will need to check if 'c' is a digit. Assume it is..
                    var d = c - '0';

                    // skip leading 0s
                    if ((d | accumulator) == 0)
                        continue;

                    // try add a digit
                    if (!TryAddNoOverflow(ref accumulator, d, sign))
                    {
                        System.Console.WriteLine("not representable " + (i));
                    }
                }
            }
        }

        // accumulate the value and ensure the sign
        // returns false when number is no longer representable in 32bit
        static bool TryAddNoOverflow(ref int value, int nextDigit, int sign)
        {
            var newValue = value * 10 + (nextDigit * sign);

            // opposite sign check for {newValue, sign}
            if ((newValue ^ sign) < 0)
            {
                return false;
            }

            value = newValue;
            return true;
        }
```